arch/arm/include/aarch32/cortex_m/wrap_func.h

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief
+ */
+
+#ifndef ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_M_WRAP_FUNC_H_
+#define ZEPHYR_ARCH_ARM_INCLUDE_AARCH32_CORTEX_M_WRAP_FUNC_H_
+
+#ifdef _ASMLANGUAGE
+
+/* nothing */
+
+#else
+
+/**
+ * @brief Macro for "sandwiching" a function call (@p name) in two other calls
+ *
+ * This macro should be called via @ref Z_ARM_WRAP_FUNC or
+ * @ref Z_ARM_WRAP_FUNC_STACK_ARGS.
+ *
+ * This macro creates the function body of an "outer" function which behaves
+ * exactly like the wrapped function (@p name), except that the preface function
+ * is called before, and the postface function afterwards.
+ *
+ * @param preface   The function to call first. Must have no parameters and no
+ *                  return value.
+ * @param name      The main function, i.e. the function to wrap. This function
+ *                  will receive the arguments, and its return value will be
+ *                  returned.
+ * @param postface  The function to call last. Must have no parameters and no
+ *                  return value.
+ * @param store_lr  The assembly instruction for storing away the LR value
+ *                  before the functions are called.
+ * @param load_lr   The assembly instruction for restoring the LR value after
+ *                  the functions have been called.
+ */
+#define Z_ARM_WRAP_FUNC_RAW(preface, name, postface, store_lr, load_lr) \
+	do { \
+		__asm(".global "#preface"; .type "#preface", %function"); \
+		__asm(".global "#name"; .type "#name", %function"); \
+		__asm(".global "#postface"; .type "#postface", %function"); \
+		\
+		__asm(store_lr); \
+		__asm("push {r0-r3}"); \
+		__asm("bl "#preface); \
+		__asm("pop {r0-r3}"); \
+		__asm("bl "#name); \
+		__asm("push {r0-r3}"); \
```

Comment:
Since the post function has no arguments, do we need to push the whole list r0-r3? I suppose r0 would hold the return value from the function, possibly r1 as well. Not sure about r3,r2.

Reply:
Yes, I found that a sometimes you need r1 as well, so I assumed it's theoretically possible need all 4, but maybe not?